interpreter.go

```diff
@@ -181,74 +92,82 @@ type callStack struct {
 	stack []*callFrame
 }
 
+func dumpCallStack(c *callStack) string {
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "<callStack calls = %d limit = %d stack:\n", c.calls, c.limit)
+	for _, callFrame := range c.stack {
+		fmt.Fprintf(&buf, "  %v\n", dumpCallFrame(callFrame))
+	}
+	buf.WriteString("\n>")
+	return buf.String()
+}
+
 func (s *callStack) top() *callFrame {
 	return s.stack[len(s.stack)-1]
-
 }
 
 func (s *callStack) pop() {
 	if s.top().isCall {
 		s.calls--
 	}
 	s.stack = s.stack[:len(s.stack)-1]
-
 }
 
+// TODO(sbarzowski) I don't get this. When we have a tail call why can't we just
+// pop the last call from stack before pushing our new thing.
```

Comment:
I can't remember.  Feel free to try other approaches and make sure there is good test coverage.

Potential cases:

1) mark a call tailstrict when it's not a tail call
2) mark a call tailstrict when it's not being called from itself
3) mark a call tailstrict when the arguments are not actually strict